From 43ea2de8282b3e2e33ff5a427fe2c3077df77b74 Mon Sep 17 00:00:00 2001
From: Christian Schwede <cschwede@redhat.com>
Date: Thu, 15 Oct 2015 08:39:54 +0000
Subject: [PATCH] Retry creating of floating IPs on failures

Currently nodepools deletes and restarts a new instance if the creation of a new
floating IP fails.  However, it might be that this is a temporary shortage of
available IPs, and in that case a lot of instances might be restarted
continuous.

This patch changes this behavior, and tries to create a new floating IP
repeatedly until it either succeeds or hits the default timeout of 600 seconds.

Change-Id: I9ba5ec717497daba64e6120da0bff2360078ca87
---

diff --git a/nodepool/provider_manager.py b/nodepool/provider_manager.py
index bd08c6f..8e7bb8c 100644
--- a/nodepool/provider_manager.py
+++ b/nodepool/provider_manager.py
@@ -506,7 +506,13 @@
         return self._waitForResource('image', image_id, timeout)
 
     def createFloatingIP(self, pool=None):
-        return self.submitTask(CreateFloatingIPTask(pool=pool))
+        for count in iterate_timeout(
+                600, "Creating floating IP on %s" % (self.provider.name)):
+            try:
+                return self.submitTask(CreateFloatingIPTask(pool=pool))
+            except novaclient.exceptions.ClientException:
+                self.log.exception('Unable to create floating IP on %s, '
+                                   'will retry' % (self.provider.name))
 
     def addFloatingIP(self, server_id, address):
         self.submitTask(AddFloatingIPTask(server=server_id,
